File: src/main/java/fr/paulevans/incidents/model/Incident.java
package fr.paulevans.incidents.model;

import fr.paulevans.incidents.annotation.ValidEnum;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "incidents")
public class Incident {

    @Id
    private String id;

    @NotEmpty
    private String title;

    @NotEmpty
    private String summary; // what the incident is about

    @NotEmpty
    private String severity;

    @NotNull
    @ValidEnum(enumClass = IncidentStatus.class)
    private String status;

    @NotEmpty
    private String createdBy; // who created the incident

    @NotNull
    private Instant createdAt; // when it was created
    private Instant updatedAt; // when it was updated

    private String resolutionNote; // optional note explaining resolution
    private Instant resolvedAt; // when it was resolved
    private List<@Valid TimelineEvent> timeline; // chronological log of events
    private List<@Valid Note> notes;
    private List<String> tags;

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class TimelineEvent {

        @Id
        private String id;

        @NotNull
        private Instant timestamp;

        @NotEmpty
        private String description; // what happened

        @NotEmpty
        private String actor; // who performed this action
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class Note {

        @Id
        private String id;

        @NotEmpty
        private String author;

        @NotEmpty
        private String note;

        @NotNull
        private Instant timestamp;
    }
}

File: frontend/src/types/incident.ts
export interface Incident {
  id: string;
  title: string;
  summary: string;
  severity: string;
  status: IncidentStatus;
  createdBy: string;
  createdAt: string;
  updatedAt?: string;
  resolutionNote?: string;
  resolvedAt?: string;
  timeline: TimelineEvent[];
  notes: Note[];
  tags: string[];
}

export interface IncidentSummary {
  id: string;
  title: string;
  summary: string;
  severity: string;
  status: IncidentStatus;
  createdBy: string;
  createdAt: string;
  resolutionNote?: string;
  resolvedAt?: string;
  tags: string[];
}

export interface TimelineEvent {
  id: string;
  timestamp: string;
  description: string;
  actor: string;
}

export interface Note {
  id: string;
  author: string;
  note: string;
  timestamp: string;
}

export type IncidentStatus = 'OPEN' | 'IN_PROGRESS' | 'RESOLVED' | 'CLOSED';

export interface CreateIncidentRequest {
  title: string;
  summary: string;
  severity: string;
  status: IncidentStatus;
  createdBy: string;
  createdAt: string;
  tags: string[];
  timeline: Omit<TimelineEvent, 'id'>[];
  notes: Omit<Note, 'id'>[];
}